src/snowflake/snowpark/dataframe.py

```diff
@@ -5516,6 +5533,8 @@ def first(
 
             # Flush the AST and encode it as part of the query.
             _, kwargs[DATAFRAME_AST_PARAMETER] = self._session._ast_batch.flush(stmt)
+            # Also pass the AST UID for query listeners
+            kwargs["dataframeAstId"] = self._ast_id
```

Comment:
discussed offline, we could make use of `df._plan.uuid` for this to remove dependency on ast id.